src/workerd/api/headers.c++

```diff
@@ -0,0 +1,728 @@
+  V("location")                                                                                    \
+  V("max-forwards")                                                                                \
+  V("proxy-authenticate")                                                                          \
+  V("proxy-authorization")                                                                         \
+  V("range")                                                                                       \
+  V("referer")                                                                                     \
+  V("refresh")                                                                                     \
+  V("retry-after")                                                                                 \
+  V("server")                                                                                      \
+  V("set-cookie")                                                                                  \
+  V("strict-transport-security")                                                                   \
+  V("transfer-encoding")                                                                           \
+  V("user-agent")                                                                                  \
+  V("vary")                                                                                        \
+  V("via")                                                                                         \
+  V("www-authenticate")
+
+// Constexpr array of lowercase common header names (must match CommonHeaderName enum order
+// and must be kept in sync with the ordinal values defined in http-over-capnp.capnp). Since
+// it is extremely unlikely that those will change often, we hardcode them here for runtime
+// efficiency.
+#define V(Name) Name##_kj,
+constexpr kj::StringPtr COMMON_HEADER_NAMES[] = {nullptr,  // 0: invalid
+  COMMON_HEADERS(V)};
+#undef V
+
+inline constexpr kj::StringPtr getCommonHeaderName(uint id) {
+  KJ_ASSERT(id > 0 && id <= MAX_COMMON_HEADER_ID, "Invalid common header ID");
+  return COMMON_HEADER_NAMES[id];
+}
+
+constexpr bool strcaseeq(kj::StringPtr a, kj::StringPtr b) {
+  if (a.size() != b.size()) return false;
+  for (size_t i = 0; i < a.size(); ++i) {
+    char ca = a[i];
+    char cb = b[i];
+    // Convert to lowercase for comparison
+    if ('A' <= ca && ca <= 'Z') ca += 32;
+    if ('A' <= cb && cb <= 'Z') cb += 32;
+    if (ca != cb) return false;
+  }
+  return true;
+}
+
+constexpr uint caseInsensitiveHash(kj::StringPtr name) {
+  uint hash = 2166136261u;
+  for (size_t i = 0; i < name.size(); ++i) {
+    char c = name[i];
+    if ('A' <= c && c <= 'Z') c += 32;
+    hash ^= static_cast<uint8_t>(c);
+    hash *= 16777619u;
+  }
+  return hash;
+}
+
+constexpr size_t HEADER_MAP_SIZE = 128;
+
+// Constexpr hash table for case-insensitive mapping of header names to their
```

Comment:
Since the entire contents of the hash table are known in advance, you should be able to twiddle the hash function and the size until you get a "perfect hash" with no collisions at all. Then you don't need to implement any probing.